codalab/apps/web/tasks.py

```diff
@@ -605,7 +608,99 @@ def __init__(self, submission, inner_exception):
 
 
 @app.task(queue='submission-updates')
-def update_submission(job_id, args, secret):
+def register_worker(worker_id, ip, cpu_count, mem_mb, harddrive_gb, gpus, vhost):
+    logger.info("Registering worker id = {}, ip = {}, cpu_count = {}, mem_mb = {}, harddrive_gb = {}, gpus = {}, queue = {}".format(
+        worker_id,
+        ip,
+        cpu_count,
+        mem_mb,
+        harddrive_gb,
+        gpus,
+        vhost,
+    ))
+
+    if vhost != '/':
+        queue = Queue.objects.get(vhost=vhost)
+    else:
+        queue = None
+
+    worker = Worker.objects.update_or_create(
+        unique_id=worker_id,
+        defaults={
+            "ip": ip,
+            "cpu_count": cpu_count,
+            "mem_mb": mem_mb,
+            "harddrive_gb": harddrive_gb,
+            "gpus": gpus,
+            "queue": queue
+        },
+    )
+
+
+@app.task(queue='submission-updates')
+def worker_job_started(worker_id, submission_secret, is_scoring):
+    assert worker_id, "WORKER_ID cannot be null, received: {}".format(worker_id)
+    logger.info("Starting job worker id = {}, submission_secret = {}, is_scoring = {}".format(
+        worker_id,
+        submission_secret,
+        is_scoring,
+    ))
+    TaskMetadata.objects.update_or_create(
+        submission=CompetitionSubmission.objects.get(secret=submission_secret),
+        is_predicting=not is_scoring,
+        is_scoring=is_scoring,
+        defaults={
+            "worker": Worker.objects.get_or_create(unique_id=worker_id)[0],
+            "start": timezone.now(),
+        }
+    )
+
+
+@app.task(queue='submission-updates')
+def worker_job_ended(worker_id, submission_secret, is_scoring):
+    assert worker_id, "WORKER_ID cannot be null, received: {}".format(worker_id)
+    logger.info("Ending job worker id = {}, submission_secret = {}, is_scoring = {}".format(
+        worker_id,
+        submission_secret,
+        is_scoring,
+    ))
+    TaskMetadata.objects.update_or_create(
+        submission=CompetitionSubmission.objects.get(secret=submission_secret),
+        is_predicting=not is_scoring,
+        is_scoring=is_scoring,
+        defaults={
+            "worker": Worker.objects.get_or_create(unique_id=worker_id)[0],
+            "end": timezone.now(),
+        }
+    )
+
+
+@app.task(queue='submission-updates')
+def worker_keep_alive(worker_id):
+    assert worker_id, "WORKER_ID cannot be null, received: {}".format(worker_id)
+    logger.info("Keep alive received for worker id = {}".format(worker_id))
+    worker = Worker.objects.get_or_create(unique_id=worker_id)[0]
```

Comment:
To keep w/ general django convention, shouldn't we use tuple unpacking here?
Or what about update_or_create instead?